File: scripts/lib/sets.mjs
/**
 * Recording which pictures a command run made.
 *
 * A `/shotall` is forty-six renders of one character and a `/photostory` is a
 * shoot in stages, but both land in the same dated Forge output folder as
 * everything else rendered that day, interleaved with unrelated work. Nothing
 * on the files says they belong together: the prompts differ by design — that
 * is the whole point of a set — and a bracket reuses one seed across five
 * frames, so neither prompt nor seed can be the key.
 *
 * So the run writes a manifest beside its pictures, and the vault reads it on
 * the next scan. See `apps/desktop/src/sets.rs`, which is the other half.
 *
 * **A file, not a call into the app.** Renders queued with `--queue` drain when
 * the GPU is free, which is deliberately the middle of the night — there may be
 * no app running to tell, and a set recorded only in a database would be lost
 * the next time the index is rebuilt. A file on disk survives both.
 *
 * **The member list is discovered, not chosen.** Locally Forge saves its own
 * copy under its own numbering (`00905-4090734728.png`) and never tells the API
 * what it called it, so the run has to go and look. It looks by seed, which the
 * render *does* return, narrowed to files that appeared while this render was
 * running — a seed reused across a bracket would otherwise match its own
 * earlier frames.
 */

import { existsSync, mkdirSync, readdirSync, readFileSync, statSync, writeFileSync } from 'node:fs'
import { basename, dirname, join } from 'node:path'

import { FORGE, IS_REMOTE, RENDER_DIR, forge } from './forge.mjs'

/** The folder a manifest goes in, beside the pictures it describes. */
export const MANIFEST_DIR = '.luma-sets'

/**
 * Parse the `--set` argument.
 *
 * `<command>/<character>/<stamp>[/<title>]` — for example
 * `shotall/alexstrasza/20260903T1431`. One argument rather than four flags, and
 * no separate "begin a run" step, because every render of a set is its own CLI
 * invocation: whatever identifies the run has to be something the caller can
 * repeat verbatim forty-six times, and a string it composed once is the only
 * thing that survives that without a state file to lose.
 *
 * The stamp is what makes two `/shotall`s of the same character on the same day
 * two sets rather than one, so it is required and not defaulted — a run that
 * forgets it would silently swallow the previous one's pictures.
 */
export function parseSet(value) {
  if (!value) return null
  const parts = String(value)
    .split('/')
    .map((part) => part.trim())
    .filter(Boolean)
  if (parts.length < 3) {
    throw new Error(
      `--set wants <command>/<character>/<stamp>, got "${value}"\n` +
        '  e.g. --set shotall/alexstrasza/20260903T1431',
    )
  }
  const [command, character, stamp, ...rest] = parts
  return {
    run: slug(`${command}-${character}-${stamp}`),
    command,
    character,
    title: rest.length > 0 ? rest.join(' / ') : null,
    createdAt: stampToMs(stamp),
  }
}

/** Lowercase, and only the characters a filename and a key can both hold. */
function slug(text) {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

/**
 * `20260903T1431` → unix ms, falling back to now.
 *
 * Read as local time, because the stamp was written by something looking at a
 * local clock and a set that lists two hours before it was shot reads as a bug.
 */
function stampToMs(stamp) {
  const match = /^(\d{4})(\d{2})(\d{2})T?(\d{2})?(\d{2})?$/.exec(stamp)
  if (!match) return Date.now()
  const [, y, mo, d, h = '0', mi = '0'] = match
  return new Date(Number(y), Number(mo) - 1, Number(d), Number(h), Number(mi)).getTime()
}

/**
 * Where Forge saves its own copies, asked once per process.
 *
 * The API reports the outdir as Forge sees it (`X:/AI/…`), which is not the
 * name the vault indexes the same tree under (`\\jebpot\devs\AI\…`). That is
 * fine and is the reason the manifest lists bare filenames: this path is only
 * used to *find* the files and to decide which folder the manifest goes in, and
 * both sides then resolve members against that folder.
 */
let outdirPromise = null
export function forgeOutdir() {
  outdirPromise ??= forge('/sdapi/v1/options')
    .then((options) => ({
      root: options?.outdir_txt2img_samples || options?.outdir_samples || null,
      // Forge files into `<outdir>/<date>/` unless this is off.
      toDirs: options?.save_to_dirs !== false,
    }))
    .catch(() => ({ root: null, toDirs: true }))
  return outdirPromise
}

/**
 * The folder this render's picture landed in, and what it is called there.
 *
 * Returns `null` when it cannot be found, which is not an error: the picture
 * exists either way, and a set that is missing a frame is better than a run
 * that stops because it could not file one.
 */
async function locate({ seed, since }) {
  // Remote Forge keeps its own copy on the rented machine, so the copy the
  // library sees is the one we filed ourselves — and we know its name exactly.
  if (IS_REMOTE) return null

  const { root, toDirs } = await forgeOutdir()
  if (!root) return null

  // `[date]` is the default directory pattern, and the only one worth guessing
  // at: a render that started before midnight can land in either day's folder,
  // so both are searched.
  const days = toDirs ? [dayFolder(since), dayFolder(Date.now())] : ['']
  for (const day of [...new Set(days)]) {
    const folder = day ? join(root, day) : root
    if (!existsSync(folder)) continue
    const found = readdirSync(folder)
      .filter((name) => new RegExp(`-${seed}\\.[a-z0-9]+$`, 'i').test(name))
      .map((name) => ({ name, at: mtime(join(folder, name)) }))
      // A bracket reuses one seed across five renders, so the seed alone names
      // several files. The one this render made is the one that appeared while
      // it was running; newest wins if two somehow share the moment.
      .filter((entry) => entry.at >= since - 2000)
      .sort((a, b) => b.at - a.at)
    if (found.length > 0) return { folder, file: found[0].name }
  }
  return null
}

const mtime = (path) => {
  try {
    return statSync(path).mtimeMs
  } catch {
    return 0
  }
}

const dayFolder = (ms) => {
  const at = new Date(ms)
  const pad = (n) => String(n).padStart(2, '0')
  return `${at.getFullYear()}-${pad(at.getMonth() + 1)}-${pad(at.getDate())}`
}

/**
 * Add one picture to a run's manifest.
 *
 * Appends after every render rather than writing once at the end, so a run that
 * is interrupted — and a forty-six render run gets interrupted — still leaves a
 * set holding everything it managed to make. Re-adding a file already listed
 * updates its label instead of duplicating it, which is what makes a retried
 * shot behave.
 *
 * Never throws. Filing is bookkeeping wrapped around the thing that actually
 * matters, and a picture that rendered must not be lost to a failure to record
 * where it went.
 */
export async function recordMember(set, { seed, since, label, destination }) {
  if (!set) return null
  try {
    const found = IS_REMOTE
      ? RENDER_DIR
        ? { folder: RENDER_DIR, file: basename(destination) }
        : null
      : await locate({ seed, since })
    if (!found) return null

    const dir = join(found.folder, MANIFEST_DIR)
    mkdirSync(dir, { recursive: true })
    const path = join(dir, `${set.run}.json`)

    const manifest = readManifest(path) ?? {
      run: set.run,
      command: set.command,
      character: set.character,
      title: set.title,
      createdAt: set.createdAt,
      members: [],
    }
    const existing = manifest.members.findIndex((member) => member.file === found.file)
    const member = { file: found.file, ...(label ? { label } : {}) }
    if (existing >= 0) manifest.members[existing] = member
    else manifest.members.push(member)

    writeFileSync(path, JSON.stringify(manifest, null, 2) + '\n')
    return path
  } catch {
    return null
  }
}

function readManifest(path) {
  try {
    const manifest = JSON.parse(readFileSync(path, 'utf8'))
    return Array.isArray(manifest?.members) ? manifest : null
  } catch {
    // A half-written manifest from a killed run is worth less than the set it
    // is about to describe: start again rather than refuse to record anything.
    return null
  }
}

/** Only for the error message, so a run says where Forge is when it cannot look. */
export const FORGE_URL = FORGE

File: apps/desktop/src/sets.rs
//! Sets: the pictures one command run made, recorded so they can be found
//! together again.
//!
//! A `/shotall` is sixteen angles of one character; a `/photostory` is a shoot
//! in stages. Both land in the same dated Forge output folder as everything
//! else rendered that day, interleaved with unrelated work, and nothing on the
//! files themselves says they belong together — the prompts differ by design,
//! and the seeds are reused across a bracket, so neither can be the key.
//!
//! So the run says so, in a file it writes beside the pictures:
//!
//! ```text
//! …/txt2img-images/2026-09-03/
//!   00905-4090734728.png
//!   00906-1590191850.png
//!   .luma-sets/
//!     shotall-alexstrasza-20260903T1431-7f3a.json
//! ```
//!
//! **A file rather than a call into the app.** The queue drains when the GPU is
//! free, which is the middle of the night by design — there may be no app
//! running to tell, and a set that only exists in a database is lost the next
//! time the index is rebuilt. A manifest on disk survives both, and is re-read
//! by every scan.
//!
//! **Members are bare filenames, resolved against the manifest's own folder.**
//! Two reasons, and either alone would be enough. Forge names its own output
//! (`00905-4090734728.png`) and never tells the API what it called it, so a run
//! cannot write paths it chose. And the two machines disagree about what the
//! path *is*: Forge reports `X:/AI/Stable Diffusion/outputs/…` while the vault
//! indexes the same tree as `\\jebpot\devs\AI\Stable Diffusion\outputs\…`. A
//! name relative to the manifest is the one form both agree on.

use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

/// The directory a run writes its manifest into, beside the pictures.
pub const MANIFEST_DIR: &str = ".luma-sets";

/// One command run, as its manifest describes it.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SetManifest {
    /// Unique across the library, and stable across re-reads: the run names
    /// itself once and appends to the same file as it goes.
    pub run: String,
    /// The command that made it — `shotall`, `photostory`.
    pub command: String,
    /// Who it is of, when the run knows. The sidebar groups by this, and a run
    /// that cannot name a character still lists under "Other".
    #[serde(default)]
    pub character: Option<String>,
    /// What to call it in a list. Absent is fine — the command and the time
    /// make a serviceable name without one.
    #[serde(default)]
    pub title: Option<String>,
    /// When the run started, unix ms.
    pub created_at: i64,
    #[serde(default)]
    pub members: Vec<SetMember>,
}

/// One picture of a run.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SetMember {
    /// The file's name in the folder above the manifest. A name, never a path:
    /// see the module note.
    pub file: String,
    /// What the run called this shot — "from below", "stage 3 — undressed".
    #[serde(default)]
    pub label: Option<String>,
}

/// Is this the manifest of a set?
///
/// By its folder, not by its name. A run names its file after itself so two
/// runs on the same day cannot collide, which means the name cannot also be a
/// marker — and a stray `.json` dropped into an output folder must not be read
/// as one.
pub fn is_manifest(path: &Path) -> bool {
    path.extension().and_then(|ext| ext.to_str()).map(str::to_lowercase).as_deref() == Some("json")
        && path
            .parent()
            .and_then(Path::file_name)
            .and_then(|name| name.to_str())
            .is_some_and(|name| name.eq_ignore_ascii_case(MANIFEST_DIR))
}

/// Read a manifest, and say where its pictures are.
///
/// Returns the manifest and the absolute path of each member, in the order the
/// run recorded them — which is the order it shot them, and the order a set
/// should be read in. A member naming anything but a plain filename is dropped
/// rather than resolved: this turns a string from a file into a path the
/// indexer will look up, and `../../` in it has no honest reading.
pub fn read_manifest(path: &Path) -> Result<(SetManifest, Vec<(PathBuf, SetMember)>), String> {
    let text = std::fs::read_to_string(path).map_err(|error| format!("{error}"))?;
    let manifest: SetManifest =
        serde_json::from_str(&text).map_err(|error| format!("unreadable manifest: {error}"))?;
    if manifest.run.trim().is_empty() {
        return Err("manifest names no run".to_string());
    }

    // `.luma-sets/<run>.json` → the folder holding the pictures.
    let Some(folder) = path.parent().and_then(Path::parent) else {
        return Err("manifest is not inside a folder".to_string());
    };

    let members = manifest
        .members
        .iter()
        .filter(|member| is_plain_filename(&member.file))
        .map(|member| (folder.join(&member.file), member.clone()))
        .collect();

    Ok((manifest, members))
}

/// Rewrite a manifest with its members in a new order.
///
/// Members are named by file, and every name given must already be in the
/// manifest: this reorders what a run recorded, it does not add to it or remove
/// from it. Anything the caller does not mention keeps its relative position at
/// the end, so a manifest that grew between the app reading it and the user
/// dragging something is not quietly truncated.
pub fn reorder_members(manifest: &SetManifest, order: &[String]) -> SetManifest {
    let mut moved: Vec<SetMember> = Vec::with_capacity(manifest.members.len());
    for name in order {
        if let Some(member) = manifest.members.iter().find(|each| &each.file == name) {
            if !moved.iter().any(|each| each.file == member.file) {
                moved.push(member.clone());
            }
        }
    }
    for member in &manifest.members {
        if !moved.iter().any(|each| each.file == member.file) {
            moved.push(member.clone());
        }
    }
    SetManifest {
        members: moved,
        ..manifest.clone()
    }
}

/// Write a manifest, atomically.
///
/// Temp file then rename, and the rename is the point rather than tidiness. The
/// app's own watcher is live on this folder, and three STATUS_HEAP_CORRUPTION
/// exits all happened while manifests were being rewritten under it — one of
/// them by a render script that rewrote its manifest forty times in an
/// afternoon. A rename is one event and can never be observed half-written; a
/// truncate-and-write is at least two and can be.
///
/// The temp file is created in the same directory so the rename stays on one
/// filesystem, and it is named to be ignored by [`is_manifest`] — otherwise the
/// watcher would read the half-written file we are trying to hide from it.
pub fn write_manifest(path: &Path, manifest: &SetManifest) -> Result<(), String> {
    let Some(dir) = path.parent() else {
        return Err("manifest has no directory".to_string());
    };
    std::fs::create_dir_all(dir).map_err(|error| format!("{error}"))?;

    let text = serde_json::to_string_pretty(manifest)
        .map_err(|error| format!("cannot serialise manifest: {error}"))?;

    // `.tmp`, not `.json`: `is_manifest` matches on the extension, so a `.json`
    // temp beside the real one would be read as a second set of the same run.
    let temp = path.with_extension("json.tmp");
    std::fs::write(&temp, text.as_bytes()).map_err(|error| format!("{error}"))?;
    std::fs::rename(&temp, path).map_err(|error| {
        let _ = std::fs::remove_file(&temp);
        format!("{error}")
    })
}

/// A name that names a file in one folder, and cannot mean anywhere else.
fn is_plain_filename(name: &str) -> bool {
    !name.is_empty()
        && !name.contains('/')
        && !name.contains('\\')
        && !name.contains(':')
        && name != "."
        && name != ".."
}

#[cfg(test)]
mod tests {
    use super::*;

    fn manifest_json() -> &'static str {
        r#"{
            "run": "shotall-alexstrasza-20260903T1431-7f3a",
            "command": "shotall",
            "character": "alexstrasza",
            "title": "Alexstrasza — every angle",
            "createdAt": 1772547060000,
            "members": [
                { "file": "00905-4090734728.png", "label": "from below" },
                { "file": "00906-1590191850.png" }
            ]
        }"#
    }

    #[test]
    fn a_manifest_is_recognised_by_its_folder_not_its_name() {
        assert!(is_manifest(Path::new("/out/2026-09-03/.luma-sets/run.json")));
        // The name is the run's, so it cannot also be the marker.
        assert!(!is_manifest(Path::new("/out/2026-09-03/run.json")));
        assert!(!is_manifest(Path::new("/out/2026-09-03/.luma-sets/notes.txt")));
        assert!(!is_manifest(Path::new("/out/.luma-sets")));
    }

    #[test]
    fn members_resolve_against_the_folder_above_the_manifest() {
        let dir = tempfile::tempdir().expect("tempdir");
        let sets = dir.path().join("2026-09-03").join(MANIFEST_DIR);
        std::fs::create_dir_all(&sets).expect("create");
        let path = sets.join("run.json");
        std::fs::write(&path, manifest_json()).expect("write");

        let (manifest, members) = read_manifest(&path).expect("readable");
        assert_eq!(manifest.command, "shotall");
        assert_eq!(manifest.character.as_deref(), Some("alexstrasza"));
        assert_eq!(members.len(), 2);
        assert_eq!(members[0].0, dir.path().join("2026-09-03").join("00905-4090734728.png"));
        assert_eq!(members[0].1.label.as_deref(), Some("from below"));
        // Order is the run's, because it is the order the pictures were shot.
        assert_eq!(members[1].0.file_name().unwrap(), "00906-1590191850.png");
    }

    #[test]
    fn a_member_cannot_name_a_file_outside_the_folder() {
        let dir = tempfile::tempdir().expect("tempdir");
        let sets = dir.path().join("2026-09-03").join(MANIFEST_DIR);
        std::fs::create_dir_all(&sets).expect("create");
        let path = sets.join("run.json");
        std::fs::write(
            &path,
            r#"{"run":"r","command":"shotall","createdAt":1,"members":[
                {"file":"../../../etc/passwd"},
                {"file":"sub/nested.png"},
                {"file":"C:/Windows/win.ini"},
                {"file":".."},
                {"file":"good.png"}
            ]}"#,
        )
        .expect("write");

        let (_, members) = read_manifest(&path).expect("readable");
        let names: Vec<_> =
            members.iter().map(|(path, _)| path.file_name().unwrap().to_owned()).collect();
        assert_eq!(names, vec!["good.png"], "only plain filenames may resolve");
    }

    #[test]
    fn reordering_moves_named_members_and_keeps_the_rest() {
        let manifest = SetManifest {
            run: "r".into(),
            command: "shotall".into(),
            character: None,
            title: None,
            created_at: 1,
            members: vec![
                SetMember { file: "a.png".into(), label: None },
                SetMember { file: "b.png".into(), label: Some("two".into()) },
                SetMember { file: "c.png".into(), label: None },
            ],
        };

        let moved = reorder_members(&manifest, &["c.png".to_string(), "a.png".to_string()]);
        let names: Vec<&str> = moved.members.iter().map(|m| m.file.as_str()).collect();
        // b was not named, so it keeps its place at the end rather than vanishing:
        // a manifest that grew since the app read it must not be truncated.
        assert_eq!(names, vec!["c.png", "a.png", "b.png"]);
        assert_eq!(moved.members[2].label.as_deref(), Some("two"));
        assert_eq!(moved.run, "r");
    }

    #[test]
    fn reordering_ignores_a_name_the_manifest_does_not_have() {
        let manifest = SetManifest {
            run: "r".into(),
            command: "shotall".into(),
            character: None,
            title: None,
            created_at: 1,
            members: vec![SetMember { file: "a.png".into(), label: None }],
        };
        let moved = reorder_members(&manifest, &["ghost.png".to_string(), "a.png".to_string()]);
        assert_eq!(moved.members.len(), 1);
        assert_eq!(moved.members[0].file, "a.png");
    }

    #[test]
    fn writing_round_trips_and_leaves_no_temp_behind() {
        let dir = std::env::temp_dir().join(format!("luma-sets-write-{}", std::process::id()));
        let manifest_dir = dir.join(MANIFEST_DIR);
        std::fs::create_dir_all(&manifest_dir).expect("dirs");
        let path = manifest_dir.join("r.json");

        let manifest = SetManifest {
            run: "r".into(),
            command: "photostory".into(),
            character: Some("Mira".into()),
            title: Some("set 042".into()),
            created_at: 7,
            members: vec![SetMember { file: "a.png".into(), label: Some("one".into()) }],
        };
        write_manifest(&path, &manifest).expect("write");

        let (back, members) = read_manifest(&path).expect("read");
        assert_eq!(back, manifest);
        assert_eq!(members.len(), 1);
        // The temp must not survive, and must never have looked like a manifest.
        assert!(!manifest_dir.join("r.json.tmp").exists());
        assert!(!is_manifest(&manifest_dir.join("r.json.tmp")));
        std::fs::remove_dir_all(&dir).ok();
    }

    #[test]
    fn a_damaged_manifest_is_an_error_not_a_panic() {
        let dir = tempfile::tempdir().expect("tempdir");
        let sets = dir.path().join(MANIFEST_DIR);
        std::fs::create_dir_all(&sets).expect("create");

        let truncated = sets.join("truncated.json");
        std::fs::write(&truncated, r#"{"run":"r","command":"shot"#).expect("write");
        assert!(read_manifest(&truncated).is_err());

        let anonymous = sets.join("anonymous.json");
        std::fs::write(&anonymous, r#"{"run":"  ","command":"shotall","createdAt":1}"#)
            .expect("write");
        assert!(read_manifest(&anonymous).is_err(), "a run with no name cannot be recorded");
    }
}
